proof/crefine/RISCV64/IpcCancel_C.thy

```diff
@@ -2934,6 +2939,105 @@ lemma reply_remove_ccorres:
     (replyRemove replyPtr tcbPtr) (Call reply_remove_'proc)"
 sorry (* FIXME RT: reply_remove_ccorres *)
 
+lemma getBlockingObject_BlockedOnReceive_return:
+  "(getBlockingObject (Structures_H.thread_state.BlockedOnReceive bo bicg ro)) = return bo"
+  unfolding getBlockingObject_def epBlocked_def by simp
+
+lemma getBlockingObject_BlockedOnSend_return:
+  "(getBlockingObject (Structures_H.thread_state.BlockedOnSend bo bib bicg bicgr biic)) = return bo"
+  unfolding getBlockingObject_def epBlocked_def by simp
+
+lemma BlockedOnReceive_replyObject_no_0:
+  assumes "no_0_obj' s"
+  assumes "valid_tcb_state' ts s"
+  assumes "ts = BlockedOnReceive bo bicg ro"
+  shows "ro \<noteq> Some 0"
+  using assms
+  by (auto simp: valid_tcb_state'_def)
```

Comment:
For these simple lemmas, declaring all the assumptions via `assumes` and then immediately using `assms` doesn't gain a lot in terms of readability. You can compress them vertically a lot, e.g.:

```
lemma BlockedOnReceive_replyObject_no_0:
  "⟦ no_0_obj' s; valid_tcb_state' ts s; ts = BlockedOnReceive bo bicg ro ⟧
   ⟹ ro ≠ Some 0"
  by (auto simp: valid_tcb_state'_def)
```

As an aside, rules tend to match better in Isabelle when their first assumption is more specific, since you usually want to make it unify with one of your goal assumptions. In this case `valid_tcb_state' ts s`  or `ts = BlockedOnReceive bo bicg ro` might be more useful to go first.

Feel free to ask us questions if it's not clear why.

Reply:
I've compressed these lemmas by preferring the `⟦ ... ⟧ ⟹ ...` syntax and reordered some assumptions based on the tip you gave.

Reply:
That looks much cleaner, thank you.